### PR TITLE
mogd optimization

### PR DESCRIPTION
**Pull request by Angryrou**
Several optimizations:

1. when `Constraint.function` is `th.nn.Module`, set `.requires_grad=False` for all parameters, and set the `function.eval()` to make sure the dropout and batch-norm will work in the eval mode instead of training model.

2. when `input_data` is `UdaoEmbedInput`, MOGD will compute the embedding at each iteration to get identical results. When the embedded has large parameters (e.g., a graph transformer model), it wastes lots of time. To solve the issue, `MOGD.embedding` caches the embedding results when the objective function is `DerivedUdaoModel`.

3. update the `run.py` to use objectives with `DerivedUdaoModel` to cache the `embedding` during MOGD

4. might need help on the test case for double check.

### CODE REVIEW

**Comment by Angryrou** [comment]
Current dependencies on/for this PR:
* `dev`
  * **PR #6** <a href="https://app.graphite.dev/github/pr/Angryrou/udao/6?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #9** <a href="https://app.graphite.dev/github/pr/Angryrou/udao/9?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
      * **PR #10** <a href="https://app.graphite.dev/github/pr/Angryrou/udao/10?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/Angryrou/udao/9?utm_source=stack-comment).

**Review by julhien — CHANGES_REQUESTED** [review]
To avoid handling specific cases in MOGD, and to make the unique computation of embeddings available in all solvers, I suggest to store the value of the embedding within the derived model and not worry about it in the solver itself.

```python

class FixedEmbeddingUdaoModel(th.nn.Module):
    def __init__(self, model: UdaoModel, embedding: Optional[th.Tensor] = None) -> None:
        super().__init__()
        self.model = model
        self.embedding = embedding

    def forward(
        self, input_data: UdaoEmbedInput
    ) -> th.Tensor:
        if embedding is None:
            self.embedding = self.embedder(input_data.embedding_input)
        return self.model(input_data, self.embedding)
```

File: udao/optimization/soo/mogd.py [hunk 1]
```diff
@@ -228,9 +228,12 @@ def _gradient_descent(
         ----------
         problem : co.SOProblem
             Single-objective optimization problem
-        input_data : Any
+        input_data : Union[UdaoInput, List, Dict]
```
Comment (julhien) [inline udao/optimization/soo/mogd.py]:
```suggestion
        input_data : Union[UdaoInput, Dict]
```

File: udao/optimization/soo/mogd.py [hunk 2]
```diff
@@ -228,9 +228,12 @@ def _gradient_descent(
...
             Input data - can have different types depending on whether
-            the input variables are processed or not
+            the input variables are processed or not.
+            - UdaoInput: the naive input
+            - List: [UdaoEmbedInput, embedding: th.Tensor]
```
Comment (julhien) [inline udao/optimization/soo/mogd.py]:
```suggestion
```

File: udao/optimization/soo/mogd.py [hunk 3]
```diff
@@ -460,7 +463,11 @@ def _processed_single_start_opt(
             input_data.features[:, grad_indices] = input_vars_subvector
             try:
                 min_loss_id, min_loss, local_best_obj = self._gradient_descent(
-                    problem, input_data, optimizer=optimizer
+                    problem,
+                    input_data
+                    if isinstance(input_data, UdaoEmbedInput)
+                    else input_data,
```
Comment (julhien) [inline udao/optimization/soo/mogd.py]:
```suggestion
                    input_data,
```

File: udao/model/model.py [hunk 4]
```diff
@@ -35,7 +35,22 @@ def __init__(self, embedder: BaseEmbedder, regressor: BaseRegressor) -> None:
...
+    def forward(
+        self, input_data: UdaoEmbedInput, embedding: Optional[th.Tensor] = None
+    ) -> th.Tensor:
+        if embedding is None:
+            embedding = self.embedder(input_data.embedding_input)
         inst_feat = input_data.features
         return self.regressor(embedding, inst_feat)
+
+
+class FixedEmbeddingUdaoModel(th.nn.Module):
```
Comment (julhien) [inline udao/model/model.py:47]:
```suggestion
class FixedEmbeddingUdaoModel(th.nn.Module):
"""
Assumes the embedding part does not change between inputs and caches it after computing it once. This is relevant for an optimization pipeline where the embedding input (e.g. a query plan) is a fixed parameter
"""
```

**Review by julhien — APPROVED** [review]
